### ISSUE TITLE
Please stop the C# auto-formatter indent based on global scope

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/roslyn#43189]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/content/problem/977720/please-stop-the-c-auto-formatter-indent-based-on-g.html)._

---
<p>The C# auto-formatter indents according to its view of the world, not the scope it’s really in. If you have some code that (for good reason) is not indented exacly to the level the auto-formatter would indent it to if it was let lose to do its thing for the whole file, it indents code within the currently edited scope completely up the walls.</p>
<p>Let’s say you have some code inside a method that the auto-formatter thinks should be at indentation level N, but you have for good reason de-indented it to be at level M. Let’s take a concrete example.</p>
<p>(using indent=2 here to keep it readable in a web browser)</p>
<pre>  void method()
  {
    ...
      <span class=hljs-keyword>if</span> (something) <span class=hljs-comment>// &lt;- manually de-indented to level M</span>
      {
        <span class=hljs-keyword>if</span> (blah1 &amp;&amp; blah2 &amp;&amp; many_more_checks)
        {
        }
      }
    ...
  }
</pre>
<p>If you now decide to break the conditional checks up into multi-line, you might f.ex. select the first <code>&amp;&amp;</code>, engage multi-select with Shift-Alt-Dot (both because it’s 100-1000 times faster than waiting for the refactoring to suggest it, and because you want to keep <code>&amp;&amp;</code> at EOL), hit Right, and finalize with Return, expecting the following</p>
<pre>  void method()
  {
    ...
      <span class=hljs-keyword>if</span> (something)
      {
        <span class=hljs-keyword>if</span> (blah1 &amp;&amp;
            blah2 &amp;&amp;
            many_more_checks)
        {
        }
      }
    ...
  }
</pre>
<p>Instead you are given</p>
<pre>  void method()
  {
    ...
      <span class=hljs-keyword>if</span> (something)
      {
        <span class=hljs-keyword>if</span> (blah1 &amp;&amp;
                blah2 &amp;&amp; <span class=hljs-comment>// &lt;- bad auto-formatter, no cookie [1]</span>
                many_more_checks)
        {
        }
      }
    ...
  }
</pre>
<p>Now you need to manually de-indent this broken indentation the auto-formatter inserted。</p>
<p>The reason for this misbehaviour is that the auto-formatter doesn’t give a rat’s about what the surrounding scope indentation is. It looks at the indentation from the method (or even file) level, and indents the code according to how <em>it</em> thinks the world should look - globally.</p>
<p>The fix is obvious; Priority #1 for the auto-formatter’s indentation choice should be to indent relative to the surrounding scope. Only if that fails could it be let out of its cage </p>
<p>Sidenote: This pain is induced for many more editing scenarios, for the same reason, but perhaps getting this specific scenario fixed could provide a starting point.</p>
<p>[1] I won’t even touch the fact it’s using indent N+1 there, instead of the proper <em>indent</em> M followed by four spaces for <em>alignment</em> (<code>sizeof(&quot;if (&quot;)</code>) – I’ve already experienced the futility of trying to explain the difference between indentation and aligment.</p>

---
#### Visual Studio Feedback System on 4/7/2020, 01:14 AM: 

<p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

---
### Original Solutions
(no solutions)

**Comment by sharwell** [comment]
I agree that the behavior observed above is incorrect. The indentation engine should use the narrowest (i.e. innermost) scope as the baseline for determining relative indentation adjustments.
